File: JavaScript/kuantum-kaos-yonetimi.js
#!/usr/bin/env node
/**
 * Kuantum Kaos Yönetimi Sistemi - JavaScript Implementation
 * Nesne Yönelimli Programlama Prensipleri:
 * - Abstract Class & Encapsulation (ES6 Classes ile)
 * - Interface Segregation (Duck Typing ile)
 * - Inheritance & Polymorphism
 */

/**
 * Interface Segregation - IKritik
 * JavaScript'te interface yoktur, ancak duck typing kullanılır
 * Kritik durumlarda acil müdahale gerektiren nesneler için
 */
class IKritik {
    acilDurumSogutmasi() {
        throw new Error("acilDurumSogutmasi() metodu implement edilmelidir!");
    }
}

/**
 * Abstract Class & Encapsulation - KuantumNesnesi
 * Tüm kuantum nesnelerinin temel sınıfı
 */
class KuantumNesnesi {
    /**
     * Constructor
     * @param {string} id - Nesne kimliği
     */
    constructor(id) {
        if (this.constructor === KuantumNesnesi) {
            throw new Error("Abstract class KuantumNesnesi doğrudan instantiate edilemez!");
        }

        // Encapsulation - Private fields (ES2022 syntax)
        this._id = id;
        this._stabilite = 50.0; // Varsayılan stabilite
    }

    // Getter ve Setter metodları (Encapsulation)
    get id() {
        return this._id;
    }

    set id(value) {
        this._id = value;
    }

    get stabilite() {
        return this._stabilite;
    }

    set stabilite(value) {
        if (value >= 0 && value <= 100) {
            this._stabilite = value;
        } else {
            throw new Error("Stabilite 0-100 arasında olmalıdır!");
        }
    }

    /**
     * Abstract method - Alt sınıflar tarafından implement edilmeli
     */
    tehlikeSeviyesi() {
        throw new Error("tehlikeSeviyesi() metodu alt sınıflarda implement edilmelidir!");
    }

    /**
     * Virtual method - Override edilebilir
     */
    durumBilgisi() {
        console.log(`Nesne ID: ${this._id}, Stabilite: ${this._stabilite}%`);
    }
}

/**
 * Inheritance & Polymorphism - Ambar sınıfı
 * KuantumNesnesi'nden türetilmiş ve IKritik interface'ini implement eder
 */
class Ambar extends KuantumNesnesi {
    /**
     * Constructor
     * @param {string} id - Ambar kimliği
     * @param {string} amir - Vardiya amiri adı
     */
    constructor(id, amir) {
        super(id);
        this._yerVardiyaAmirsiniz = amir;
    }

    // Getter ve Setter
    get yerVardiyaAmirsiniz() {
        return this._yerVardiyaAmirsiniz;
    }

    set yerVardiyaAmirsiniz(value) {
        this._yerVardiyaAmirsiniz = value;
    }

    /**
     * Override - Tehlike seviyesini değerlendirir ve rapor eder
     */
    tehlikeSeviyesi() {
        if (this.stabilite < 30) {
            console.log(`UYARI: ${this.id} ambarı kritik seviyede! Stabilite: ${this.stabilite}%`);
        } else if (this.stabilite < 60) {
            console.log(`DİKKAT: ${this.id} ambarı orta risk seviyesinde. Stabilite: ${this.stabilite}%`);
        } else {
            console.log(`GÜVENLİ: ${this.id} ambarı güvenli seviyede. Stabilite: ${this.stabilite}%`);
        }
    }

    /**
     * IKritik interface implementation - Acil durum soğutması
     */
    acilDurumSogutmasi() {
        console.log(`${this.id} ambarı için acil durum soğutması başlatıldı!`);
        this.stabilite = Math.min(100, this.stabilite + 20);
    }

    /**
     * Override - Genişletilmiş durum bilgisi
     */
    durumBilgisi() {
        super.durumBilgisi();
        console.log(`Vardiya Amiri: ${this._yerVardiyaAmirsiniz}`);
    }
}

/**
 * Inheritance & Polymorphism - Metot sınıfı
 * KuantumNesnesi'nden türetilmiş sınıf
 */
class Metot extends KuantumNesnesi {
    /**
     * Constructor
     * @param {string} id - Metot kimliği
     * @param {string} analizTipi - Analiz türü
     */
    constructor(id, analizTipi) {
        super(id);
        this._analizTipi = analizTipi;
        this._durumBilgisiAktif = true;
    }

    // Getter ve Setter metodları
    get analizTipi() {
        return this._analizTipi;
    }

    set analizTipi(value) {
        this._analizTipi = value;
    }

    get durumBilgisiAktif() {
        return this._durumBilgisiAktif;
    }

    set durumBilgisiAktif(value) {
        this._durumBilgisiAktif = value;
    }

    /**
     * Override - Tehlike seviyesini değerlendirir ve rapor eder
     */
    tehlikeSeviyesi() {
        if (this.stabilite < 20) {
            console.log(`KRİTİK: ${this.id} metodu çökme riski altında! Stabilite: ${this.stabilite}%`);
        } else if (this.stabilite < 50) {
            console.log(`UYARI: ${this.id} metodu kararsız. Stabilite: ${this.stabilite}%`);
        } else {
            console.log(`STABIL: ${this.id} metodu normal çalışıyor. Stabilite: ${this.stabilite}%`);
        }
    }

    /**
     * Override - Genişletilmiş durum bilgisi
     */
    durumBilgisi() {
        super.durumBilgisi();
        console.log(`Analiz Tipi: ${this._analizTipi}, Durum Bilgisi Aktif: ${this._durumBilgisiAktif}`);
    }
}

/**
 * Ana program fonksiyonu - Menü döngüsü
 */
async function main() {
    console.log("=== KUANTUM KAOS YÖNETİMİ SİSTEMİ ===");
    console.log("JavaScript Implementation\n");

    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });

    const question = (query) => new Promise((resolve) => readline.question(query, resolve));

    const getValidStability = async (prompt) => {
        while (true) {
            const value = parseFloat(await question(prompt));
            if (!isNaN(value) && value >= 0 && value <= 100) {
                return value;
            }
            console.log("Hatalı giriş! 0-100 arasında bir sayı giriniz.");
        }
    };

    const getValidMenuChoice = async (prompt, max) => {
        while (true) {
            const value = parseInt(await question(prompt));
            if (!isNaN(value) && value >= 1 && value <= max) {
                return value;
            }
            console.log(`Hatalı giriş! 1-${max} arasında bir sayı giriniz.`);
        }
    };

    try {
        const kuantumNesneleri = [];

        // Ana döngü (Main Loop)
        while (true) {
            console.log("\n" + "=".repeat(50));
            console.log("KUANTUM AMBARI KONTROL PANELİ");
            console.log("=".repeat(50));
            console.log("1. Yeni Nesne Ekle (Rastgele Veri/Kararlık Madde/Anti Madde üretir)");
            console.log("2. Nesneleri Listele (Durum Raporu)");
            console.log("3. Tehlikeyi Analiz Et (ID integersı)");
            console.log("4. Acil Durum Soğutması Yap (Sadece İKritik olanlar için)");
            console.log("5. Çıkış");

            const secim = await getValidMenuChoice("Seçiminiz (1-5): ", 5);

            if (secim === 1) {
                console.log("\n=== YENİ NESNE EKLE ===");
                console.log("1. Ambar");
                console.log("2. Metot");
                const tip = await getValidMenuChoice("Nesne tipi seçiniz (1-2): ", 2);

                const nesneId = await question("ID giriniz: ");
                const stabilite = await getValidStability("Stabilite giriniz (0-100): ");

                if (tip === 1) {
                    const amir = await question("Vardiya Amiri giriniz: ");
                    const nesne = new Ambar(nesneId, amir);
                    nesne.stabilite = stabilite;
                    kuantumNesneleri.push(nesne);
                    console.log(`✓ Ambar ${nesneId} başarıyla eklendi!`);
                } else {
                    const analizTipi = await question("Analiz Tipi giriniz: ");
                    const nesne = new Metot(nesneId, analizTipi);
                    nesne.stabilite = stabilite;
                    kuantumNesneleri.push(nesne);
                    console.log(`✓ Metot ${nesneId} başarıyla eklendi!`);
                }
            } else if (secim === 2) {
                console.log("\n=== NESNE LİSTESİ (DURUM RAPORU) ===");
                if (kuantumNesneleri.length === 0) {
                    console.log("Henüz hiç nesne eklenmemiş.");
                } else {
                    kuantumNesneleri.forEach((nesne, i) => {
                        console.log(`\n${i + 1}. Nesne:`);
                        nesne.durumBilgisi();
                    });
                }
            } else if (secim === 3) {
                console.log("\n=== TEHLİKE ANALİZİ ===");
                if (kuantumNesneleri.length === 0) {
                    console.log("Henüz hiç nesne eklenmemiş.");
                } else {
                    kuantumNesneleri.forEach(nesne => {
                        nesne.tehlikeSeviyesi();
                        console.log("---");
                    });
                }
            } else if (secim === 4) {
                console.log("\n=== ACİL DURUM SOĞUTMASI ===");
                if (kuantumNesneleri.length === 0) {
                    console.log("Henüz hiç nesne eklenmemiş.");
                } else {
                    const kritikNesneler = kuantumNesneleri.filter(n => typeof n.acilDurumSogutmasi === 'function');
                    if (kritikNesneler.length === 0) {
                        console.log("Soğutma yapılabilecek nesne bulunamadı.");
                    } else {
                        kritikNesneler.forEach(nesne => {
                            nesne.acilDurumSogutmasi();
                            console.log("---");
                        });
                    }
                }
            } else if (secim === 5) {
                console.log("\n=== SİSTEM RAPORU ===");
                const kritikSayisi = kuantumNesneleri.filter(nesne => nesne.stabilite < 30).length;
                const guvenliSayisi = kuantumNesneleri.filter(nesne => nesne.stabilite >= 60).length;

                console.log(`Toplam Nesne: ${kuantumNesneleri.length}`);
                console.log(`Kritik Seviye: ${kritikSayisi}`);
                console.log(`Güvenli Seviye: ${guvenliSayisi}`);
                console.log(`Orta Risk Seviye: ${kuantumNesneleri.length - kritikSayisi - guvenliSayisi}`);
                console.log("\nProgram sonlandırılıyor...");
                break;
            }
        }
    } catch (error) {
        console.error("Bir hata oluştu:", error);
    } finally {
        readline.close();
    }
}

// Program başlatma
if (typeof require !== 'undefined' && require.main === module) {
    main();
}

// Export for module usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        KuantumNesnesi,
        Ambar,
        Metot,
        IKritik,
        main
    };
}

File: Python/kuantum_kaos_yonetimi.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Kuantum Kaos Yönetimi Sistemi - Python Implementation
Nesne Yönelimli Programlama Prensipleri:
- Abstract Class & Encapsulation
- Interface Segregation  
- Inheritance & Polymorphism
"""

from abc import ABC, abstractmethod
from typing import List, Protocol
import random


class IKritik(Protocol):
    """
    Interface Segregation - IKritik protokolü
    Kritik durumlarda acil müdahale gerektiren nesneler için
    """
    def acil_durum_sogutmasi(self) -> None:
        """Acil durum soğutması başlatır"""
        pass


class KuantumNesnesi(ABC):
    """
    Abstract Class & Encapsulation - KuantumNesnesi
    Tüm kuantum nesnelerinin temel sınıfı
    """
    
    def __init__(self, id_: str):
        """
        Constructor
        Args:
            id_: Nesne kimliği
        """
        self._id = id_  # Encapsulation - private attribute
        self._stabilite = 50.0  # Varsayılan stabilite
    
    @property
    def id(self) -> str:
        """ID getter (Encapsulation)"""
        return self._id
    
    @id.setter
    def id(self, value: str) -> None:
        """ID setter (Encapsulation)"""
        self._id = value
    
    @property
    def stabilite(self) -> float:
        """Stabilite getter (Encapsulation)"""
        return self._stabilite
    
    @stabilite.setter
    def stabilite(self, value: float) -> None:
        """
        Stabilite setter (Encapsulation)
        Args:
            value: 0-100 arasında stabilite değeri
        Raises:
            ValueError: Stabilite değeri geçersizse
        """
        if 0 <= value <= 100:
            self._stabilite = value
        else:
            raise ValueError("Stabilite 0-100 arasında olmalıdır!")
    
    @abstractmethod
    def tehlike_seviyesi(self) -> None:
        """Abstract method - Alt sınıflar tarafından implement edilmeli"""
        pass
    
    def durum_bilgisi(self) -> None:
        """Virtual method - Override edilebilir"""
        print(f"Nesne ID: {self._id}, Stabilite: {self._stabilite}%")


class Ambar(KuantumNesnesi):
    """
    Inheritance & Polymorphism - Ambar sınıfı
    KuantumNesnesi'nden türetilmiş ve IKritik interface'ini implement eder
    """
    
    def __init__(self, id_: str, amir: str):
        """
        Constructor
        Args:
            id_: Ambar kimliği
            amir: Vardiya amiri adı
        """
        super().__init__(id_)
        self._yer_vardiya_amirsiniz = amir
    
    @property
    def yer_vardiya_amirsiniz(self) -> str:
        """Vardiya amiri getter"""
        return self._yer_vardiya_amirsiniz
    
    @yer_vardiya_amirsiniz.setter
    def yer_vardiya_amirsiniz(self, value: str) -> None:
        """Vardiya amiri setter"""
        self._yer_vardiya_amirsiniz = value
    
    def tehlike_seviyesi(self) -> None:
        """Tehlike seviyesini değerlendirir ve rapor eder"""
        if self.stabilite < 30:
            print(f"UYARI: {self.id} ambarı kritik seviyede! Stabilite: {self.stabilite}%")
        elif self.stabilite < 60:
            print(f"DİKKAT: {self.id} ambarı orta risk seviyesinde. Stabilite: {self.stabilite}%")
        else:
            print(f"GÜVENLİ: {self.id} ambarı güvenli seviyede. Stabilite: {self.stabilite}%")
    
    def acil_durum_sogutmasi(self) -> None:
        """IKritik interface implementation - Acil durum soğutması"""
        print(f"{self.id} ambarı için acil durum soğutması başlatıldı!")
        self.stabilite = min(100, self.stabilite + 20)
    
    def durum_bilgisi(self) -> None:
        """Override - Genişletilmiş durum bilgisi"""
        super().durum_bilgisi()
        print(f"Vardiya Amiri: {self._yer_vardiya_amirsiniz}")


class Metot(KuantumNesnesi):
    """
    Inheritance & Polymorphism - Metot sınıfı
    KuantumNesnesi'nden türetilmiş sınıf
    """
    
    def __init__(self, id_: str, analiz_tipi: str):
        """
        Constructor
        Args:
            id_: Metot kimliği
            analiz_tipi: Analiz türü
        """
        super().__init__(id_)
        self._analiz_tipi = analiz_tipi
        self._durum_bilgisi_aktif = True
    
    @property
    def analiz_tipi(self) -> str:
        """Analiz tipi getter"""
        return self._analiz_tipi
    
    @analiz_tipi.setter
    def analiz_tipi(self, value: str) -> None:
        """Analiz tipi setter"""
        self._analiz_tipi = value
    
    @property
    def durum_bilgisi_aktif(self) -> bool:
        """Durum bilgisi aktiflik durumu getter"""
        return self._durum_bilgisi_aktif
    
    @durum_bilgisi_aktif.setter
    def durum_bilgisi_aktif(self, value: bool) -> None:
        """Durum bilgisi aktiflik durumu setter"""
        self._durum_bilgisi_aktif = value
    
    def tehlike_seviyesi(self) -> None:
        """Tehlike seviyesini değerlendirir ve rapor eder"""
        if self.stabilite < 20:
            print(f"KRİTİK: {self.id} metodu çökme riski altında! Stabilite: {self.stabilite}%")
        elif self.stabilite < 50:
            print(f"UYARI: {self.id} metodu kararsız. Stabilite: {self.stabilite}%")
        else:
            print(f"STABIL: {self.id} metodu normal çalışıyor. Stabilite: {self.stabilite}%")
    
    def durum_bilgisi(self) -> None:
        """Override - Genişletilmiş durum bilgisi"""
        super().durum_bilgisi()
        print(f"Analiz Tipi: {self._analiz_tipi}, Durum Bilgisi Aktif: {self._durum_bilgisi_aktif}")


def get_valid_stability(prompt: str) -> float:
    """Kullanıcıdan geçerli bir stabilite değeri alır"""
    while True:
        try:
            value = float(input(prompt))
            if 0 <= value <= 100:
                return value
            print("Hatalı giriş! 0-100 arasında bir sayı giriniz.")
        except ValueError:
            print("Hatalı giriş! Lütfen sayısal bir değer giriniz.")

def get_valid_menu_choice(prompt: str, max_choice: int) -> int:
    """Kullanıcıdan geçerli bir menü seçimi alır"""
    while True:
        try:
            value = int(input(prompt))
            if 1 <= value <= max_choice:
                return value
            print(f"Hatalı giriş! 1-{max_choice} arasında bir sayı giriniz.")
        except ValueError:
            print("Hatalı giriş! Lütfen sayısal bir değer giriniz.")

def yeni_nesne_ekle(kuantum_nesneleri: List[KuantumNesnesi]) -> None:
    """Yeni nesne ekleme fonksiyonu"""
    print("\n=== YENİ NESNE EKLE ===")
    print("1. Ambar")
    print("2. Metot")
    tip = get_valid_menu_choice("Nesne tipi seçiniz (1-2): ", 2)
    
    nesne_id = input("ID giriniz: ")
    stabilite = get_valid_stability("Stabilite giriniz (0-100): ")
    
    if tip == 1:
        amir = input("Vardiya Amiri giriniz: ")
        nesne = Ambar(nesne_id, amir)
        nesne.stabilite = stabilite
        kuantum_nesneleri.append(nesne)
        print(f"✓ Ambar {nesne_id} başarıyla eklendi!")
    else:
        analiz_tipi = input("Analiz Tipi giriniz: ")
        nesne = Metot(nesne_id, analiz_tipi)
        nesne.stabilite = stabilite
        kuantum_nesneleri.append(nesne)
        print(f"✓ Metot {nesne_id} başarıyla eklendi!")

def nesneleri_listele(kuantum_nesneleri: List[KuantumNesnesi]) -> None:
    """Nesneleri listeleme fonksiyonu"""
    print("\n=== NESNE LİSTESİ (DURUM RAPORU) ===")
    if not kuantum_nesneleri:
        print("Henüz hiç nesne eklenmemiş.")
        return
    
    for i, nesne in enumerate(kuantum_nesneleri, 1):
        print(f"\n{i}. Nesne:")
        nesne.durum_bilgisi()

def tehlikeyi_analiz_et(kuantum_nesneleri: List[KuantumNesnesi]) -> None:
    """Tehlike analizi fonksiyonu"""
    print("\n=== TEHLİKE ANALİZİ ===")
    if not kuantum_nesneleri:
        print("Henüz hiç nesne eklenmemiş.")
        return
    
    for nesne in kuantum_nesneleri:
        nesne.tehlike_seviyesi()
        print("---")

def acil_durum_sogutmasi_yap(kuantum_nesneleri: List[KuantumNesnesi]) -> None:
    """Acil durum soğutması fonksiyonu"""
    print("\n=== ACİL DURUM SOĞUTMASI ===")
    if not kuantum_nesneleri:
        print("Henüz hiç nesne eklenmemiş.")
        return
    
    kritik_nesneler = [n for n in kuantum_nesneleri if hasattr(n, 'acil_durum_sogutmasi')]
    if not kritik_nesneler:
        print("Soğutma yapılabilecek nesne bulunamadı.")
        return
    
    for nesne in kritik_nesneler:
        nesne.acil_durum_sogutmasi()
        print("---")

def main():
    """Ana program fonksiyonu - Menü döngüsü"""
    print("=== KUANTUM KAOS YÖNETİMİ SİSTEMİ ===")
    print("Python Implementation\n")
    
    # Polymorphism - Farklı türdeki nesneleri aynı listede saklama
    kuantum_nesneleri: List[KuantumNesnesi] = []
    
    # Ana döngü (Main Loop)
    while True:
        print("\n" + "="*50)
        print("KUANTUM AMBARI KONTROL PANELİ")
        print("="*50)
        print("1. Yeni Nesne Ekle (Rastgele Veri/Kararlık Madde/Anti Madde üretir)")
        print("2. Nesneleri Listele (Durum Raporu)")
        print("3. Tehlikeyi Analiz Et (ID integersı)")
        print("4. Acil Durum Soğutması Yap (Sadece İKritik olanlar için)")
        print("5. Çıkış")
        
        secim = get_valid_menu_choice("Seçiminiz (1-5): ", 5)
        
        if secim == 1:
            yeni_nesne_ekle(kuantum_nesneleri)
        elif secim == 2:
            nesneleri_listele(kuantum_nesneleri)
        elif secim == 3:
            tehlikeyi_analiz_et(kuantum_nesneleri)
        elif secim == 4:
            acil_durum_sogutmasi_yap(kuantum_nesneleri)
        elif secim == 5:
            print("\n=== SİSTEM RAPORU ===")
            kritik_sayisi = sum(1 for nesne in kuantum_nesneleri if nesne.stabilite < 30)
            guvenli_sayisi = sum(1 for nesne in kuantum_nesneleri if nesne.stabilite >= 60)
            
            print(f"Toplam Nesne: {len(kuantum_nesneleri)}")
            print(f"Kritik Seviye: {kritik_sayisi}")
            print(f"Güvenli Seviye: {guvenli_sayisi}")
            print(f"Orta Risk Seviye: {len(kuantum_nesneleri) - kritik_sayisi - guvenli_sayisi}")
            print("\nProgram sonlandırılıyor...")
            break


if __name__ == "__main__":
    main()
